File: leetcode-04/solutions/cpp/0015.cpp
class Solution {
 public:
  vector<vector<int>> threeSum(vector<int>& nums) {
    if (nums.size() < 3) return {};

    vector<vector<int>> ans;

    sort(nums.begin(), nums.end());

    for (int i = 0; i < nums.size() - 2; i++) {
      if (i == 0 || (nums[i] != nums[i - 1])) {
        int l = i + 1;
        int r = nums.size() - 1;
        int target = -nums[i];
        while (l < r) {
          if (nums[l] + nums[r] == target) {
            ans.push_back({nums[i], nums[l], nums[r]});
            while (l < r && nums[l] == nums[l + 1]) l++;
            while (l < r && nums[r] == nums[r - 1]) r--;
            l++;
            r--;
          } else if (nums[l] + nums[r] < target) {
            l++;
          } else {
            r--;
          }
        }
      }
    }

    return ans;
  }
};

File: leetcode-04/solutions/python3/0015.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []

        ans = []

        nums.sort()

        for i in range(len(nums) - 2):
            if i == 0 or i > 0 and nums[i] != nums[i - 1]:
                l = i + 1
                r = len(nums) - 1
                target = -nums[i]
                while l < r:
                    if nums[l] + nums[r] == target:
                        ans.append((nums[i], nums[l], nums[r]))
                        while l < r and nums[l] == nums[l + 1]:
                            l += 1
                        while l < r and nums[r] == nums[r - 1]:
                            r -= 1
                        l += 1
                        r -= 1
                    elif nums[l] + nums[r] < target:
                        l += 1
                    else:
                        r -= 1

        return ans
